// better-auth/dist/plugins/one-tap/index.mjs
import { parseUserOutput } from "../../db/schema.mjs";
import { setSessionCookie } from "../../cookies/index.mjs";
import { handleOAuthUserInfo } from "../../oauth2/link-account.mjs";
import { APIError } from "../../api/index.mjs";
import { PACKAGE_VERSION } from "../../version.mjs";
import { toBoolean } from "../../utils/boolean.mjs";
import { createAuthEndpoint } from "@better-auth/core/api";
import * as z from "zod";
import { createRemoteJWKSet, jwtVerify } from "jose";
//#region src/plugins/one-tap/index.ts
const oneTapCallbackBodySchema = z.object({ idToken: z.string().meta({ description: "Google ID token, which the client obtains from the One Tap API" }) });
const oneTap = (options) => ({
	id: "one-tap",
	version: PACKAGE_VERSION,
	endpoints: { oneTapCallback: createAuthEndpoint("/one-tap/callback", {
		method: "POST",
		body: oneTapCallbackBodySchema,
		metadata: { openapi: {
			summary: "One tap callback",
			description: "Use this endpoint to authenticate with Google One Tap",
			responses: {
				200: {
					description: "Successful response",
					content: { "application/json": { schema: {
						type: "object",
						properties: {
							session: { $ref: "#/components/schemas/Session" },
							user: { $ref: "#/components/schemas/User" }
						}
					} } }
				},
				400: { description: "Invalid token" }
			}
		} }
	}, async (ctx) => {
		const { idToken } = ctx.body;
		let payload;
		try {
			const JWKS = createRemoteJWKSet(new URL("https://www.googleapis.com/oauth2/v3/certs"));
			const googleProvider = typeof ctx.context.options.socialProviders?.google === "function" ? await ctx.context.options.socialProviders?.google() : ctx.context.options.socialProviders?.google;
			const { payload: verifiedPayload } = await jwtVerify(idToken, JWKS, {
				issuer: ["https://accounts.google.com", "accounts.google.com"],
				audience: options?.clientId || googleProvider?.clientId
			});
			payload = verifiedPayload;
		} catch {
			throw new APIError("BAD_REQUEST", { message: "invalid id token" });
		}
		const { email: rawEmail, email_verified, name, picture, sub } = payload;
		if (!rawEmail) return ctx.json({ error: "Email not available in token" });
		const result = await handleOAuthUserInfo(ctx, {
			userInfo: {
				id: sub,
				email: rawEmail.toLowerCase(),
				emailVerified: typeof email_verified === "boolean" ? email_verified : toBoolean(email_verified),
				name: name ?? "",
				image: picture
			},
			account: {
				providerId: "google",
				accountId: sub,
				idToken,
				scope: "openid,profile,email"
			},
			disableSignUp: options?.disableSignup
		});
		if (result.error) throw new APIError("UNAUTHORIZED", { message: result.error });
		await setSessionCookie(ctx, result.data);
		return ctx.json({
			token: result.data.session.token,
			user: parseUserOutput(ctx.context.options, result.data.user)
		});
	}) },
	options
});
//#endregion
export { oneTap };
